### PR TITLE
Remove decodeField methods and directly generate decodeMessage.

### PR DESCRIPTION
**Pull request by thomasvl**

### CODE REVIEW

File: Sources/SwiftProtobuf/Message.swift [hunk 1]
```diff
@@ -60,30 +60,6 @@ public protocol Message: CustomDebugStringConvertible {
...
-  ///
-  /// - Parameters:
-  ///   - decoder: a `Decoder`; the `Message` will call the method
-  ///     corresponding to the type of this field.
-  ///   - protoFieldNumber: the number of the field to decode.
-  /// - Throws: an error on failure or type mismatch.
-  mutating func decodeField<D: Decoder>(decoder: inout D, fieldNumber: Int) throws
-
   /// Decode all of the fields from the given decoder.
   ///
```
Comment (tbkka) [inline Sources/SwiftProtobuf/Message.swift]: Please update this comment (it still refers to `decodeField`)

**Comment by tbkka** [comment]
I'm curious:  were you able to measure any performance difference from this?

File: Sources/SwiftProtobuf/Message.swift (hunk 1 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/Message.swift]: New commit up with updated docs.

**Comment by thomasvl** [comment]
> I'm curious: were you able to measure any performance difference from this?

I'll try to do it this morning before pushing.

Looks like it might be time to update the perf harness to support a "mix" of fields on a message.  And also update the to support skipping C++/Instruments, and display the results of previous run vs. new run.  Since changes like this are more about just how it may or maybe have changed the ObjC speeds and not so much in comparison to the C++.

**Comment by thomasvl** [comment]
On my laptop, perfwise, the difference is in the noise.  Some of the fields seemed faster some slower.  So I'm gonna blame that all on noise in the system.  But the binaries sizes were smaller were slightly smaller in all cases, which makes sense.

File: Sources/SwiftProtobuf/Message.swift [hunk 2]
```diff
@@ -62,12 +62,25 @@ public protocol Message: CustomDebugStringConvertible {
 
   /// Decode all of the fields from the given decoder.
   ///
-  /// This is generally a simple loop that repeatedly gets the next
-  /// field number from `decoder.nextFieldNumber()` and
-  /// then invokes `decodeField` above.
+  /// This is a simple loop that repeatedly gets the next field number
+  /// from `decoder.nextFieldNumber()` and then usees the number returned
```
Comment (allevato) [inline Sources/SwiftProtobuf/Message.swift]: s/usees/uses/

File: Sources/SwiftProtobuf/Message.swift [hunk 3]
```diff
@@ -62,12 +62,25 @@ public protocol Message: CustomDebugStringConvertible {
...
+  /// and the type information from the original .proto file to  decide
```
Comment (allevato) [inline Sources/SwiftProtobuf/Message.swift]: delete extra space between "to  decide"

File: Sources/SwiftProtobuf/Message.swift [hunk 4]
```diff
@@ -62,12 +62,25 @@ public protocol Message: CustomDebugStringConvertible {
...
+  /// what type of data should be decoded for that field.  The corresponding
+  /// method on the decoder is then called to get the field value.
   ///
-  /// If you're not implementing a custom encoding format, you probably
-  /// shouldn't call this.
+  /// This is the core method used by the deserialization machinery. It is
+  /// `public` to enable users to implement their own encoding formats; it
```
Comment (allevato) [inline Sources/SwiftProtobuf/Message.swift]: s/implement their own encoding formats/implement their own encoding formats by conforming to `Decoder`/

File: Sources/SwiftProtobuf/Message.swift [hunk 5]
```diff
@@ -62,12 +62,25 @@ public protocol Message: CustomDebugStringConvertible {
...
+  /// should not be called otherwise.
+  ///
+  /// Note that this is not specific to protobuf encoding; formats that use
```
Comment (allevato) [inline Sources/SwiftProtobuf/Message.swift]: s/protobuf encoding/binary encodng/

File: Sources/SwiftProtobuf/Message.swift [hunk 6]
```diff
@@ -62,12 +62,25 @@ public protocol Message: CustomDebugStringConvertible {
...
+  /// textual identifiers translate those to fieldNumbers and also go through
```
Comment (allevato) [inline Sources/SwiftProtobuf/Message.swift]: s/fieldNumbers/field numbers/

File: Sources/SwiftProtobuf/Message.swift [hunk 7]
```diff
@@ -62,12 +62,25 @@ public protocol Message: CustomDebugStringConvertible {
...
+  /// this to decode messages.
+  ///
+  /// - Parameters:
+  ///   - decoder: a `Decoder`; the `Message` will call the method
+  ///     corresponding to the type of this field.
+  /// - Throws: an error on failure or type mismatch.  The error types
```
Comment (allevato) [inline Sources/SwiftProtobuf/Message.swift]: s/The error types will come from the `Decoder`./The type of error thrown depends on which decoder is used./

File: Sources/SwiftProtobuf/Message.swift (hunk 2 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/Message.swift]: Done

File: Sources/SwiftProtobuf/Message.swift (hunk 4 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/Message.swift]: Done

File: Sources/SwiftProtobuf/Message.swift (hunk 5 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/Message.swift]: Done

File: Sources/SwiftProtobuf/Message.swift (hunk 7 above)
Comment (thomasvl) [inline Sources/SwiftProtobuf/Message.swift]: Done

**Pull request merged**